### PR TITLE
Add confirmations for creation/removal of snap accounts

### PR DESCRIPTION
**Pull request by owencraston**
## Explanation

Progresses https://github.com/MetaMask/accounts-planning/issues/14
Designs: https://www.figma.com/file/XzaR4r04pMXIE8ft1T2hmJ/Add%2Fremove-snap-accounts%2Fkeyring-snaps-%2313-and-%2314?type=design&node-id=604-2586&mode=design

- this PR works in conjunction with changes in the [eth-snap-keyring](https://github.com/MetaMask/eth-snap-keyring/pull/99) to intercept account additions from a snap and get approval from the user.
- This PR heavily leverages existing patterns with the approval controller/confirmations flow and for that reason, differs slightly from the original designs.
    - Since we are waiting on the [accounts controller PR](https://github.com/MetaMask/metamask-extension/pull/20344) to support custom names, I have removed the TextField from the creation screen.
    - I have also opted to use the out of the box approval success/error screens with custom messages instead of the designs that @raulmono provided above. These standardized success/error screens are almost the same as the designs but lack the description and translations from the design. It is not currently possible to have custom success/error screens without a massive overhaul to the confirmations flow which would effect many teams. If changes are needed we should work with @matthewwalsh0 to make those changes upstream at a later date.

## Screenshots/Screencaps

Screenshot 2023-09-15 at 2 46 30 PM
Screenshot 2023-09-15 at 4 57 09 PM
Screenshot 2023-09-15 at 4 57 17 PM
Screenshot 2023-09-20 at 3 54 27 PM
Screenshot 2023-09-20 at 3 54 34 PM

### Before

https://github.com/MetaMask/metamask-extension/assets/22918444/465e7833-6ed1-4c38-b618-81723ed54928

#### Happy Path

https://github.com/MetaMask/metamask-extension/assets/22918444/049566e1-a121-4803-9166-1f26de2eba6f

#### Error case

https://github.com/MetaMask/metamask-extension/assets/22918444/fb628db4-4781-4bab-9efd-b755e011ff26

## Manual Testing Steps
This change requires the manual installing of the snaps rpc methods handlers and building metamask flask. Here are the steps.

#### Setup
1. checkout this branch and run `yarn start --build-type flask`
2. in your browser, load the extension by clicking `Load unpacked` in the extension settings and using the `dist/chrome/` (or whatever browser you are using) as the source
3. create your account and login

#### Testing the snap account creation
1. open this test snap: https://metamask.github.io/snap-simple-keyring/latest/
4. connect the dapp
5. click create account. Custom account names are not supported at this time
Screenshot 2023-09-20 at 4 13 31 PM
6. click create
7. a new account should be created on the snap
8. open metamask and click the accounts dropdown, a new snap account should be added

#### Testing the denial flow
1. open this test snap: https://metamask.github.io/snap-simple-keyring/latest/
2. connect the dapp
3. click create account
Screenshot 2023-09-20 at 4 13 31 PM

4. A modal should open on extension
5. click cancel
9. the modal should close and the dapp should throw an error `User denied account creation`
Screenshot 2023-09-20 at 4 14 25 PM

#### Testing the error case
- luckily for us, the snap simple keyring had a bug in version 0.2.1 that we can use to test this flow. To test the error flow, use the following steps
1. open this test snap: https://metamask.github.io/snap-simple-keyring/0.2.1/
2. connect dapp (uninstall and reinstall if you were using a different version before)
3. click create account
4. the account creation modal should pop up
5. click create, an error should be shown with the message `Method not supported: event:accountCreated`
6. create another account
10. click cancel, the modal should close and the dapp should throw an error `User denied account addition`

#### Testing account removal
1. follow the previous mentioned steps to create 2 or more snap accounts with this [snap](https://metamask.github.io/snap-simple-keyring/latest/)
2. copy the account `id` from the side bar and paste it into the remove account section of the snap
3. click remove account
Screenshot 2023-09-20 at 4 17 09 PM

4. A modal should show up indicating what you are trying to do
5. click the `share` icon beside the public address and it should open that account on etherscan. It should work with different networks as well.
6. click `remove`
7. a success screen should be shown and the snap should update with the state and the account no longer visible in the side bar
8. open the extension and verify that the account is no longer in the account list
9. repeat steps 1 and 2
10. this time click cancel
11. the extension should close and an error on the snap should appear with the text `User denied account removal`
Screenshot 2023-09-20 at 4 20 03 PM

12. verify that the account was not removed from the extension or the snap

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [x] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [x] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by owencraston** [comment]
I have read the CLA Document and I hereby sign the CLA

File: ui/components/multichain/account-list-menu/account-list-menu.js
Comment (owencraston) [inline ui/components/multichain/account-list-menu/account-list-menu.js:288]: Temporary for testing the UI

File: package.json
Comment (owencraston) [inline package.json:248]: This should be changed once a new version of the `eth-snap-keyring` with [this change is released](https://github.com/MetaMask/eth-snap-keyring/pull/99)

File: app/scripts/background.js
Comment (gantunesr) [inline app/scripts/background.js:806]: Why this one is not wrapped in the other fence?

```suggestion
          ///: END:ONLY_INCLUDE_IN
          ///: BEGIN:ONLY_INCLUDE_IN(keyring-snaps)
          case 'snap_manageAccounts:confirmation':
            controller.approvalController.accept(id, false);
            break;
```

File: ui/components/app/metamask-template-renderer/safe-component-list.js
Comment (gantunesr) [inline ui/components/app/metamask-template-renderer/safe-component-list.js:23]: Similar comment regarding using `snaps` fences and not `keyring-snaps`

**Review by gantunesr — COMMENTED** [review]
I gave it a quick look and the PR looks good! But I noticed some fences not matching the ones we use for this feature. Is that correct?

Also, can you add E2E testing to cover the test cases you specified in the PR?

File: app/scripts/background.js
Comment (matthewwalsh0) [inline app/scripts/background.js:804]: We try and maintain all approval types used across both clients in the `ApprovalType` enum defined in `@metamask/controller-utils` in the core repository.

I don't want to block this work but it would be great if we could create the necessary core PR and then make a subsequent PR to use the new enum value here.

File: app/scripts/metamask-controller.js
Comment (matthewwalsh0) [inline app/scripts/metamask-controller.js:858]: Was this logic not going to go in the `rpc-methods` package in the snaps repository?

If it's staying here, I'd always vote to minimise the size of the `MetaMaskController` where possible, so could we create a separate file to hold this logic such as `app/scripts/lib/snaps-keyring.ts`?

File: app/scripts/metamask-controller.js
Comment (matthewwalsh0) [inline app/scripts/metamask-controller.js:866]: This property is only needed if we want the code that accepts the approval request to wait for a result from this code here so we can indicate if the process being accepted was successful.

If so, there is a `resultCallbacks` property in the returned object that we need to call `success` or `error` on.

File: app/scripts/metamask-controller.js
Comment (matthewwalsh0) [inline app/scripts/metamask-controller.js:875]: We shouldn't need to use translation keys here as the extension background now supports translation also using the `t` function from `app/scripts/translate.ts`. It's exactly the same API as the frontend.

File: app/scripts/metamask-controller.js
Comment (matthewwalsh0) [inline app/scripts/metamask-controller.js:860]: As we're starting a flow, it's best practice to use a `try finally` block to ensure it gets ended no matter what otherwise we have an infinite spinner in the notification.

The approval request created below for example will throw if rejected which, looking at the code, I think would cause this scenario.

File: app/scripts/metamask-controller.js
Comment (matthewwalsh0) [inline app/scripts/metamask-controller.js:2063]: Are these hooks still needed if we're adding the logic above, or are we adding similar flows in the extension and the `rpc-methods` package?

File: shared/constants/app.ts
Comment (matthewwalsh0) [inline shared/constants/app.ts:81]: As mentioned in another comment, we try and maintain all the approval types in the enum inside `@metamask/controller-utils`.

This makes sense for now to avoid duplication but could be removed in a subsequent PR.

File: ui/pages/confirmation/confirmation.js
Comment (matthewwalsh0) [inline ui/pages/confirmation/confirmation.js:421]: Rather than adding custom logic here in the generic component, the `success` and `error` methods support an optional `header` property which lets you define any additional components you want to put in the header. This should work fine with the `SnapAuthorishipHeader` also.

```
{
	  message: `**${snapName}** is now available to use.`,
	  header: [{ key: 'snap', element: 'SnapAuthorship', props: { snapId } }],
}
```

File: ui/pages/confirmation/templates/create-snap-account.js
Comment (matthewwalsh0) [inline ui/pages/confirmation/templates/create-snap-account.js:17]: Did we not need the input support to pass through the account name?

File: ui/pages/confirmation/util.ts
Comment (matthewwalsh0) [inline ui/pages/confirmation/util.ts:52]: As mentioned in the `MetaMaskController`, we can translate in the background before we call the `success` method meaning we shouldn't need this translation support here as the burden is currently on the caller to translate.

File: ui/pages/confirmation/templates/create-snap-account.js
Comment (matthewwalsh0) [inline ui/pages/confirmation/templates/create-snap-account.js:1]: We try and add unit tests for each template just for stability and to help coverage, usually just a snapshot test to ensure the visual is what we expect. The same would apply for the remove template also.

File: app/scripts/metamask-controller.js
Comment (matthewwalsh0) [reply app/scripts/metamask-controller.js:858]: It would also support much easier unit testing which I'd recommend for this amount of logic.

**Comment by matthewwalsh0** [comment]
This looks great @owencraston! Thanks for adopting all the best practice standards with the approval flows, templates, and result pages!

File: ui/pages/confirmation/confirmation.js
Comment (owencraston) [inline ui/pages/confirmation/confirmation.js:225]: Note to self: This needs to be wrapped but the build flags dont play well with the nested comments.

File: app/scripts/metamask-controller.js
Comment (owencraston) [reply app/scripts/metamask-controller.js:875]: Ah this is perfect! Thank you

File: app/scripts/metamask-controller.js
Comment (owencraston) [reply app/scripts/metamask-controller.js]: Good catch. These are no longer used and I'll remove them.

File: app/scripts/metamask-controller.js
Comment (owencraston) [reply app/scripts/metamask-controller.js:858]: We opted to add callbacks into the [eth-snap-keyring](https://github.com/MetaMask/eth-snap-keyring/pull/99) and keep the approval logic in the extension so that we can reuse those callbacks on mobile as well as get translations.

Comment (owencraston) [reply app/scripts/metamask-controller.js:858]: I'll try and extract them out

File: ui/pages/confirmation/confirmation.js
Comment (owencraston) [reply ui/pages/confirmation/confirmation.js]: I tried to get this working but with no success. I messaged you some of my issues and also outlined the code I wrote along with the issues in this [draft PR](https://github.com/MetaMask/metamask-extension/pull/20995). For now I will move on.

File: ui/pages/confirmation/templates/create-snap-account.js
Comment (owencraston) [reply ui/pages/confirmation/templates/create-snap-account.js:17]: We are blocked on using the input meaningfully by [this PR](https://github.com/MetaMask/metamask-extension/pull/20344) so I removed it for now.

File: ui/pages/confirmation/util.ts
Comment (owencraston) [reply ui/pages/confirmation/util.ts]: Great call. removed!

File: app/scripts/metamask-controller.js
Comment (owencraston) [reply app/scripts/metamask-controller.js]: Good idea. I have added this in the latest commit.

File: app/scripts/metamask-controller.js
Comment (owencraston) [reply app/scripts/metamask-controller.js:860]: 👍  resolved in [this commit](https://github.com/MetaMask/metamask-extension/pull/20684/commits/a06f752e23a61a4ec3cc8060c965df1e4f25c83b)

File: test/e2e/snaps/test-create-snap-account.spec.js
Comment (DDDDDanica) [inline test/e2e/snaps/test-create-snap-account.spec.js:36]: here could also apply `clickElementSafe` if we need to scroll, wdyt?

File: test/e2e/snaps/test-create-snap-account.spec.js
Comment (DDDDDanica) [inline test/e2e/snaps/test-create-snap-account.spec.js:95]: Question: There was `delayStep` in `switchToWindowWithTitle`, was it really necessary to have a delay here?

File: test/e2e/snaps/test-create-snap-account.spec.js
Comment (DDDDDanica) [inline test/e2e/snaps/test-create-snap-account.spec.js:96]: Good steps of testing, would be more helpful to have some simple documentation inline to specify what we are testing /expecting

File: ui/pages/confirmation/templates/create-snap-account.test.js
Comment (DDDDDanica) [inline ui/pages/confirmation/templates/create-snap-account.test.js:54]: Suggestion: If you use `....textContent()).toMatchInlineSnapshot`, will make it easier when content changes and to find the text

File: ui/pages/create-snap-account/components/create-snap-account-content.tsx
Comment (DDDDDanica) [inline ui/pages/create-snap-account/components/create-snap-account-content.tsx:42]: Here i would suggest including `content` in the testid to indicate which file it belongs to.

File: test/e2e/snaps/test-create-snap-account.spec.js
Comment (owencraston) [reply test/e2e/snaps/test-create-snap-account.spec.js:95]: I tested without and it seemed to work perfectly fine without it so I will remove it. Thank you!

File: test/e2e/snaps/test-create-snap-account.spec.js
Comment (owencraston) [reply test/e2e/snaps/test-create-snap-account.spec.js:30]: After adding the `await driver.clickElementSafe('#connectButton');` I realized that the `findClickableElements` was not needed so I removed it.

File: app/_locales/en/messages.json
Comment (owencraston) [inline app/_locales/en/messages.json:5551]: My linter is failing on this file for some reason but is outputting no errors. Anyone know why?

File: test/e2e/snaps/test-create-snap-account.spec.js
Comment (DDDDDanica) [inline test/e2e/snaps/test-create-snap-account.spec.js:11]: Can you also make the name of test `Create Snap Account` to be consistent :)

File: app/scripts/background.js
Comment (owencraston) [reply app/scripts/background.js]: Made an issue to track this work: https://github.com/MetaMask/accounts-planning/issues/70

File: shared/constants/app.ts
Comment (owencraston) [reply shared/constants/app.ts:81]: Made an issue to track this work: https://github.com/MetaMask/accounts-planning/issues/70

**Comment by HowardBraham** [comment]
**Test summary:**
- Create Account Denial ✅
- Create Account ✅
- Remove Account Denial ✅
- Remove Account ❌ The account stays around and still exists in the Snap, but isn't registered in the Extension.  I don't know if this is a problem in the scope of this PR, but it's definitely a problem, or at least a weird unexplained user experience.
- The error case -- didn't test

File: test/e2e/accounts/test-snap-accounts.spec.js
Comment (DDDDDanica) [inline test/e2e/accounts/test-snap-accounts.spec.js:328]: Same as above, can use a const for the title :)

File: ui/pages/confirmation/confirmation.js
Comment (owencraston) [reply ui/pages/confirmation/confirmation.js]: New PR to migrate to this better pattern: https://github.com/MetaMask/metamask-extension/pull/21076/files

File: test/e2e/accounts/test-create-snap-account.spec.js
Comment (owencraston) [reply test/e2e/accounts/test-create-snap-account.spec.js:85]: I specifically made the opposite decision for the reasons outlined in Mark Stacey's post [here](https://consensys.slack.com/archives/GTQAGKY5V/p1686238528522549)


```
I'm noticing a common pattern in our e2e tests that I wanted to call attention to: find an element, then assert what its contents are.
This pattern inherently introduces a race condition, and should generally be avoided.
e.g. Instead of this:
const element = await driver.findElement('[data-testid="foo"]');
assert(await element.getText(), 'Bar')
Do this:
await driver.findElement({
  css: '[data-testid="foo"]',
  text: 'Bar',
});
This accomplishes the same thing, except that it's faster and doesn't have a race condition.
```

File: test/e2e/accounts/test-create-snap-account.spec.js
Comment (owencraston) [reply test/e2e/accounts/test-create-snap-account.spec.js:90]: see above comment

**Review by matthewwalsh0 — DISMISSED** [review]

File: test/e2e/accounts/test-create-snap-account.spec.js
Comment (owencraston) [reply test/e2e/accounts/test-create-snap-account.spec.js:50]: due to this title potentially changing in the future and its single use for testing, I am going to keep it in this file.

**Comment by owencraston** [comment]
@HowardBraham I think this is more of an issue in the snap state not being updated but once you refresh it is no longer there. I think we can address this change in the snaps code at a later date.
https://github.com/MetaMask/metamask-extension/assets/22918444/198e106d-0aa4-48dc-81c8-b7cd0c3268e3

**Comment by owencraston** [comment]
@metamaskbot update-policies

File: development/verify-locale-strings.js
Comment (owencraston) [inline development/verify-locale-strings.js:190]: Since we now have translations from the background scripts (which I added in my snap-keyring.ts file), we need to also check these paths to verify locale use.

**Review by HowardBraham — DISMISSED** [review]

**Review by matthewwalsh0 — DISMISSED** [review]

**Review by gantunesr — DISMISSED** [review]
lgtm

**Comment by gantunesr** [comment]
I didn't realize the E2E were failing before approving

**Review by matthewwalsh0 — DISMISSED** [review]

**Review by gantunesr — DISMISSED** [review]

**Comment by owencraston** [comment]
@gantunesr E2E is passing locally
Screenshot 2023-09-28 at 9 37 18 AM

**Review by matthewwalsh0 — APPROVED** [review]

File: app/scripts/background.js
Comment (hmalik88) [inline app/scripts/background.js:809]: I would prefer these approval types to live in the above enum

File: ui/pages/create-snap-account/create-snap-account.stories.tsx
Comment (hmalik88) [inline ui/pages/create-snap-account/create-snap-account.stories.tsx:4]: why is this needed?

File: ui/pages/remove-snap-account/components/remove-snap-account-content.stories.tsx
Comment (hmalik88) [inline ui/pages/remove-snap-account/components/remove-snap-account-content.stories.tsx:4]: Same as above, these comments haven't been needed before in storybook files.

File: ui/pages/remove-snap-account/remove-snap-account.stories.tsx
Comment (hmalik88) [inline ui/pages/remove-snap-account/remove-snap-account.stories.tsx:4]: Same as above

File: app/scripts/background.js
Comment (owencraston) [reply app/scripts/background.js:809]: @hmalik88 I agree, I have made an [issue here](https://github.com/MetaMask/accounts-planning/issues/70) to add them to the `ApprovalType` enum and migrate all of this custom types. We will get to this after this merges.

File: ui/pages/create-snap-account/create-snap-account.stories.tsx
Comment (owencraston) [reply ui/pages/create-snap-account/create-snap-account.stories.tsx:4]: Most of the storybook files are plain js which have less strict/different lint rules. When this line was not present the linter would fail. I believe this is the standard pattern with storybook files so I left the implementation as is and added this lint ignore. 
Screenshot 2023-09-28 at 10 45 53 AM

**Review by hmalik88 — APPROVED** [review]

**Comment by plasmacorral** [comment]
Tested commit fbb05c1599bd774012712cac7851733d6d2aff1a on Mac Sonoma 14.0 with Chrome 117 (primarily) and then some firefox 117.  Also an add and remove on Windows 11 with Firefox 117.

- Stacking multiple create/remove requests
- Trying to create/remove while extension is locked
- Creation and removal in both sync/async mode
- Rejecting creation/removal request at MM is honored in extension (but not SSK 0.2.3)
- Removing an account from the dapp that was never approved to be added to extension in the first place
- create/remove requests under poor network conditions

Observations that we can explore further:

1. Account address presented in removal confirmation is not mixed case/checksummed version - can address this in snap keyring controller in a later change
2. Able to create out of sync conditions between dapp and extension account list
        - Created but not approved accounts are presented in SSK v0.2.3, even though they do not appear in extension account list
        - Removed but not approved accounts remain present in extension, but are absent in SSK v0.2.3

The out of sync conditions could result in funds loss for a user.  This also suggests that (at least in SSK v0.2.3) that the `get accounts` and `list accounts` methods are pulling from the dapp state rather than actually from extension, as accounts created but not approved appear in the account list and removed but not approved accounts appear to be absent even though they remain in the extension account list.

**Pull request merged**
